### PR TITLE
Global planner orientation

### PR DESCRIPTION
**Pull request by DLu**
Create a variety of options for outputting orientations from the global planner. 

By default, each point will have no orientation, which is the current behavior. 

The "Forward" mode orients the robot toward the next point on the path. 

The "Interpolation" mode interpolates between the first and last pose. 

The "Mixture" model points forward most of the time until the last straightaway, whereupon it starts interpolating to its final position.

### CODE REVIEW

File: global_planner/include/global_planner/orientation_filter.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/*********************************************************************
+ * Author: David V. Lu!!
```
Comment (mikeferguson) [inline global_planner/include/global_planner/orientation_filter.h:2]: This file needs a BSD license header added

File: global_planner/src/orientation_filter.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
+
```
Comment (mikeferguson) [inline global_planner/src/orientation_filter.cpp:1]: Needs a license header

File: global_planner/cfg/GlobalPlanner.cfg [hunk 3]
```diff
@@ -10,4 +10,14 @@ gen.add("neutral_cost", int_t,   0, "Neutral Cost",  50, 1, 255)
 gen.add("cost_factor", double_t, 0, "Factor to multiply each cost from costmap by", 3.0, 0.01, 5.0)
 gen.add("publish_potential", bool_t, 0, "Publish Potential Costmap", True)
 
+orientation_enum = gen.enum([ 
+    gen.const("None",        int_t, 0, "No orientations added except goal orientation"),
+    gen.const("Forward",     int_t, 1, "Orientations point to the next point on the path"),
+    gen.const("Interpolate", int_t, 2, "Orientations are a linear blend of start and goal pose"),
+    gen.const("Mixture",     int_t, 3, "Forward orientation until last straightaway"),
```
Comment (mikeferguson) [inline global_planner/cfg/GlobalPlanner.cfg:17]: Is there a better name than "Mixture"? (or is this based off some paper where that name was used -- if so, we should include the link to said paper)

File: global_planner/include/global_planner/orientation_filter.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ *********************************************************************/
+#ifndef _ORIENTATION_FILTER_H
+#define _ORIENTATION_FILTER_H
```
Comment (mikeferguson) [inline global_planner/include/global_planner/orientation_filter.h:5]: These really should be GLOBAL_PLANNER_ORIENTATION_FILTER_H to comply with ROS style guide.

File: global_planner/src/orientation_filter.cpp [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+            for(int i=0;i<n-1;i++){
+                pointToNext(path, i);
+            }
+            
+            int i=n-3;
+            double last = getYaw(path[i]);
+            while( i>0 ){
+                double new_angle = getYaw(path[i-1]);
+                double diff = fabs(angles::shortest_angular_distance(new_angle, last));
+                // ROS_INFO("%f %f (%d)", new_angle, diff, i);
```
Comment (mikeferguson) [inline global_planner/src/orientation_filter.cpp:42]: suggest either remove these, or use ROS_DEBUG_NAMED()

File: global_planner/src/orientation_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                if( diff>0.35)
+                    break;
+                else
+                    i--;
+            }
+            // ROS_INFO("%f %f", path[i].pose.position.x, path[i].pose.position.y);
```
Comment (mikeferguson) [inline global_planner/src/orientation_filter.cpp:48]: remove or use ROS_DEBUG_NAMED()

File: global_planner/include/global_planner/planner_core.h [hunk 7]
```diff
@@ -185,6 +186,7 @@ class GlobalPlanner : public nav_core::BaseGlobalPlanner {
         PotentialCalculator* p_calc_;
         Expander* planner_;
         Traceback* path_maker_;
+        OrientationFilter* o_filter_;
```
Comment (mikeferguson) [inline global_planner/include/global_planner/planner_core.h:189]: This is only used in like 4 places -- so I would recommend making the name a bit more verbose (orientation_filter_) so people aren't having to look back and search for what the heck it is. (in the future, p_calc_ is another great variable name to update)

**Comment by mikeferguson** [comment]
Sorry to take so long getting to this -- there are a couple issues I noted above with naming, logging and licensing (especially licensing -- those have to be fixed before merging). I would also suggest compressing this to a single commit since the intermediates don't appear to have much value in the history.

File: global_planner/cfg/GlobalPlanner.cfg (hunk 3 above)
Comment (DLu) [inline global_planner/cfg/GlobalPlanner.cfg:17]: No, I made it up.

**Comment by DLu** [comment]
Closed in favor of #309.

**Comment by bxc237** [comment]
Hi, I tried the "Interpolation" mode on my omni-wheel robot with DWA local planner. It seems that the mode has no effect. The robot would still move to the goal first, then turn in place to obey the goal orientation (just like mode "None"). I am using kinetic version. Any advise or guidance would be greatly appreciated.

**Comment by reinzor** [comment]
The `orientation filter` updates the orientation of the poses in the global plan. This does not necessarily mean that the local planner is using these orientation. If you are using the `dwa_local_planner` for example, this is not the case.

**Comment by bxc237** [comment]
@reinzor Thanks for the explanation. Then which local planner should I choose to work with the orientation filter?

**Pull request closed without merging**